File: LISTAS ENLAZADAS/LinkedList.java
import java.util.Scanner;

class Node {
    int data;
    Node next;
    
    public Node(int data) {
        this.data = data;
        this.next = null;
    }
}

public class LinkedList {
    private Node head;
    private Scanner scanner;
    
    public LinkedList() {
        head = null;
        scanner = new Scanner(System.in);
    }
    
    private void clearScreen() {
        try {
            if (System.getProperty("os.name").contains("Windows")) {
                new ProcessBuilder("cmd", "/c", "cls").inheritIO().start().waitFor();
            } else {
                System.out.print("\033[H\033[2J");
                System.out.flush();
            }
        } catch (Exception e) {
            System.out.println("\n".repeat(50)); // Fallback
        }
    }
    
    public void beginset() {
        clearScreen();
        System.out.println("=== INSERTAR AL INICIO ===");
        System.out.print("Ingrese el valor del nodo: ");
        int item = scanner.nextInt();
        
        Node newNode = new Node(item);
        newNode.next = head;
        head = newNode;
        System.out.println("Nodo insertado al inicio correctamente");
    }
    
    public void lastinsert() {
        clearScreen();
        System.out.println("=== INSERTAR AL FINAL ===");
        System.out.print("Ingrese el valor del nodo: ");
        int item = scanner.nextInt();
        
        Node newNode = new Node(item);
        
        if (head == null) {
            head = newNode;
            System.out.println("Nodo insertado al final (lista estaba vacia)");
        } else {
            Node temp = head;
            while (temp.next != null) {
                temp = temp.next;
            }
            temp.next = newNode;
            System.out.println("Nodo insertado al final correctamente");
        }
    }
    
    public void randominsert() {
        clearScreen();
        System.out.println("=== INSERTAR EN POSICION ESPECIFICA ===");
        System.out.print("Ingrese el valor del elemento: ");
        int item = scanner.nextInt();
        System.out.print("Ingrese la posicion despues de la cual desea insertar: ");
        int loc = scanner.nextInt();
        
        Node newNode = new Node(item);
        
        if (head == null) {
            System.out.println("La lista esta vacia. Insertando al inicio...");
            head = newNode;
        } else {
            Node temp = head;
            for (int i = 0; i < loc; i++) {
                if (temp == null) {
                    System.out.println("No se puede insertar en esa posicion");
                    return;
                }
                temp = temp.next;
            }
            
            if (temp == null) {
                System.out.println("No se puede insertar en esa posicion");
                return;
            }
            
            newNode.next = temp.next;
            temp.next = newNode;
            System.out.println("Nodo insertado correctamente en la posicion " + (loc + 1));
        }
    }
    
    public void beginDelete() {
        clearScreen();
        System.out.println("=== ELIMINAR DEL INICIO ===");
        if (head == null) {
            System.out.println("La lista esta vacia");
        } else {
            head = head.next;
            System.out.println("Nodo eliminado del inicio correctamente");
        }
    }
    
    public void lastDelete() {
        clearScreen();
        System.out.println("=== ELIMINAR DEL FINAL ===");
        if (head == null) {
            System.out.println("La lista esta vacia");
        } else if (head.next == null) {
            head = null;
            System.out.println("Unico nodo eliminado");
        } else {
            Node temp = head;
            Node prev = null;
            while (temp.next != null) {
                prev = temp;
                temp = temp.next;
            }
            prev.next = null;
            System.out.println("Nodo eliminado del final correctamente");
        }
    }
    
    public void randomDelete() {
        clearScreen();
        System.out.println("=== ELIMINAR NODO EN POSICION ESPECIFICA ===");
        System.out.print("Ingrese la posicion del nodo que desea eliminar: ");
        int loc = scanner.nextInt();
        
        if (head == null) {
            System.out.println("La lista esta vacia");
            return;
        }
        
        if (loc == 0) {
            head = head.next;
            System.out.println("Nodo eliminado de la posicion " + loc);
            return;
        }
        
        Node temp = head;
        Node prev = null;
        for (int i = 0; i < loc; i++) {
            if (temp == null) {
                System.out.println("No se puede eliminar - posicion fuera de rango");
                return;
            }
            prev = temp;
            temp = temp.next;
        }
        
        if (temp == null) {
            System.out.println("No se puede eliminar - posicion fuera de rango");
            return;
        }
        
        prev.next = temp.next;
        System.out.println("Nodo eliminado de la posicion " + loc);
    }
    
    public void search() {
        clearScreen();
        System.out.println("=== BUSCAR ELEMENTO ===");
        if (head == null) {
            System.out.println("Lista vacia");
            return;
        }
        
        System.out.print("Ingrese el elemento que desea buscar: ");
        int item = scanner.nextInt();
        
        Node temp = head;
        int i = 0;
        boolean found = false;
        
        while (temp != null) {
            if (temp.data == item) {
                System.out.println("Elemento encontrado en la posicion " + (i + 1));
                found = true;
                break;
            }
            i++;
            temp = temp.next;
        }
        
        if (!found) {
            System.out.println("Elemento no encontrado");
        }
    }
    
    public void display() {
        clearScreen();
        System.out.println("=== MOSTRAR LISTA ===");
        if (head == null) {
            System.out.println("La lista esta vacia");
        } else {
            Node temp = head;
            System.out.print("Elementos de la lista: ");
            while (temp != null) {
                System.out.print(temp.data + " ");
                temp = temp.next;
            }
            System.out.println();
        }
    }
    
    public void menu() {
        while (true) {
            clearScreen();
            System.out.println("\n\n**Menu Principal");
            System.out.println("\nElija una opcion del siguiente menu:");
            System.out.println("1. Insertar al inicio");
            System.out.println("2. Insertar al final");
            System.out.println("3. Insertar en posicion aleatoria");
            System.out.println("4. Eliminar del inicio");
            System.out.println("5. Eliminar del final");
            System.out.println("6. Eliminar nodo despues de una posicion");
            System.out.println("7. Buscar elemento");
            System.out.println("8. Mostrar lista");
            System.out.println("9. Salir");
            
            System.out.print("\nIngrese su eleccion: ");
            int choice = scanner.nextInt();
            
            switch (choice) {
                case 1:
                    beginset();
                    break;
                case 2:
                    lastinsert();
                    break;
                case 3:
                    randominsert();
                    break;
                case 4:
                    beginDelete();
                    break;
                case 5:
                    lastDelete();
                    break;
                case 6:
                    randomDelete();
                    break;
                case 7:
                    search();
                    break;
                case 8:
                    display();
                    break;
                case 9:
                    clearScreen();
                    System.out.println("Saliendo del programa...");
                    scanner.close();
                    return;
                default:
                    clearScreen();
                    System.out.println("Por favor ingrese una opcion valida...");
                    break;
            }
            
            if (choice != 9) {
                System.out.println("\n\nPresione Enter para continuar...");
                scanner.nextLine(); // Limpiar buffer
                scanner.nextLine(); // Esperar Enter
            }
        }
    }
    
    public static void main(String[] args) {
        LinkedList lista = new LinkedList();
        lista.menu();
    }
}

File: LISTAS ENLAZADAS/LinkedList.js
const readline = require('readline');

class Node {
    constructor(data) {
        this.data = data;
        this.next = null;
    }
}

class LinkedList {
    constructor() {
        this.head = null;
        this.rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout
        });
    }

    clearScreen() {
        console.clear();
    }

    async beginset() {
        this.clearScreen();
        console.log("=== INSERTAR AL INICIO ===");
        const item = await this.question("Ingrese el valor del nodo: ");
        
        const newNode = new Node(parseInt(item));
        newNode.next = this.head;
        this.head = newNode;
        console.log("Nodo insertado al inicio correctamente");
    }

    async lastinsert() {
        this.clearScreen();
        console.log("=== INSERTAR AL FINAL ===");
        const item = await this.question("Ingrese el valor del nodo: ");
        
        const newNode = new Node(parseInt(item));
        
        if (this.head === null) {
            this.head = newNode;
            console.log("Nodo insertado al final (lista estaba vacia)");
        } else {
            let temp = this.head;
            while (temp.next !== null) {
                temp = temp.next;
            }
            temp.next = newNode;
            console.log("Nodo insertado al final correctamente");
        }
    }

    async randominsert() {
        this.clearScreen();
        console.log("=== INSERTAR EN POSICION ESPECIFICA ===");
        const item = await this.question("Ingrese el valor del elemento: ");
        const loc = await this.question("Ingrese la posicion despues de la cual desea insertar: ");
        
        const newNode = new Node(parseInt(item));
        
        if (this.head === null) {
            console.log("La lista esta vacia. Insertando al inicio...");
            this.head = newNode;
        } else {
            let temp = this.head;
            for (let i = 0; i < parseInt(loc); i++) {
                if (temp === null) {
                    console.log("No se puede insertar en esa posicion");
                    return;
                }
                temp = temp.next;
            }
            
            if (temp === null) {
                console.log("No se puede insertar en esa posicion");
                return;
            }
            
            newNode.next = temp.next;
            temp.next = newNode;
            console.log(`Nodo insertado correctamente en la posicion ${parseInt(loc) + 1}`);
        }
    }

    async beginDelete() {
        this.clearScreen();
        console.log("=== ELIMINAR DEL INICIO ===");
        if (this.head === null) {
            console.log("La lista esta vacia");
        } else {
            this.head = this.head.next;
            console.log("Nodo eliminado del inicio correctamente");
        }
    }

    async lastDelete() {
        this.clearScreen();
        console.log("=== ELIMINAR DEL FINAL ===");
        if (this.head === null) {
            console.log("La lista esta vacia");
        } else if (this.head.next === null) {
            this.head = null;
            console.log("Unico nodo eliminado");
        } else {
            let temp = this.head;
            let prev = null;
            while (temp.next !== null) {
                prev = temp;
                temp = temp.next;
            }
            prev.next = null;
            console.log("Nodo eliminado del final correctamente");
        }
    }

    async randomDelete() {
        this.clearScreen();
        console.log("=== ELIMINAR NODO EN POSICION ESPECIFICA ===");
        const loc = await this.question("Ingrese la posicion del nodo que desea eliminar: ");
        
        if (this.head === null) {
            console.log("La lista esta vacia");
            return;
        }
        
        if (parseInt(loc) === 0) {
            this.head = this.head.next;
            console.log(`Nodo eliminado de la posicion ${loc}`);
            return;
        }
        
        let temp = this.head;
        let prev = null;
        for (let i = 0; i < parseInt(loc); i++) {
            if (temp === null) {
                console.log("No se puede eliminar - posicion fuera de rango");
                return;
            }
            prev = temp;
            temp = temp.next;
        }
        
        if (temp === null) {
            console.log("No se puede eliminar - posicion fuera de rango");
            return;
        }
        
        prev.next = temp.next;
        console.log(`Nodo eliminado de la posicion ${loc}`);
    }

    async search() {
        this.clearScreen();
        console.log("=== BUSCAR ELEMENTO ===");
        if (this.head === null) {
            console.log("Lista vacia");
            return;
        }
        
        const item = await this.question("Ingrese el elemento que desea buscar: ");
        
        let temp = this.head;
        let i = 0;
        let found = false;
        
        while (temp !== null) {
            if (temp.data === parseInt(item)) {
                console.log(`Elemento encontrado en la posicion ${i + 1}`);
                found = true;
                break;
            }
            i++;
            temp = temp.next;
        }
        
        if (!found) {
            console.log("Elemento no encontrado");
        }
    }

    async display() {
        this.clearScreen();
        console.log("=== MOSTRAR LISTA ===");
        if (this.head === null) {
            console.log("La lista esta vacia");
        } else {
            let temp = this.head;
            process.stdout.write("Elementos de la lista: ");
            while (temp !== null) {
                process.stdout.write(temp.data + " ");
                temp = temp.next;
            }
            console.log();
        }
    }

    question(prompt) {
        return new Promise((resolve) => {
            this.rl.question(prompt, resolve);
        });
    }

    async waitForEnter() {
        await this.question("\n\nPresione Enter para continuar...");
    }

    async menu() {
        while (true) {
            this.clearScreen();
            console.log("\n\n**Menu Principal");
            console.log("\nElija una opcion del siguiente menu:");
            console.log("1. Insertar al inicio");
            console.log("2. Insertar al final");
            console.log("3. Insertar en posicion aleatoria");
            console.log("4. Eliminar del inicio");
            console.log("5. Eliminar del final");
            console.log("6. Eliminar nodo despues de una posicion");
            console.log("7. Buscar elemento");
            console.log("8. Mostrar lista");
            console.log("9. Salir");
            
            const choice = await this.question("\nIngrese su eleccion: ");
            
            switch (parseInt(choice)) {
                case 1:
                    await this.beginset();
                    break;
                case 2:
                    await this.lastinsert();
                    break;
                case 3:
                    await this.randominsert();
                    break;
                case 4:
                    await this.beginDelete();
                    break;
                case 5:
                    await this.lastDelete();
                    break;
                case 6:
                    await this.randomDelete();
                    break;
                case 7:
                    await this.search();
                    break;
                case 8:
                    await this.display();
                    break;
                case 9:
                    this.clearScreen();
                    console.log("Saliendo del programa...");
                    this.rl.close();
                    return;
                default:
                    this.clearScreen();
                    console.log("Por favor ingrese una opcion valida...");
                    break;
            }
            
            if (parseInt(choice) !== 9) {
                await this.waitForEnter();
            }
        }
    }
}

// Ejecutar el programa
const lista = new LinkedList();
lista.menu().catch(console.error);